dali/pipeline/operators/fused/crop_mirror_normalize.cc

```diff
@@ -14,52 +14,188 @@
+                        R"code(Output data type.)code", DALI_FLOAT)
+        .AddOptionalArg("output_layout",
+                        R"code(Output tensor data layout)code", DALI_NCHW)
+        .AddOptionalArg("pad_output",
+                        R"code(Whether to pad the output to number of channels being multiple of 4.)code",
+                        false)
+        .AddOptionalArg("crop_pos_x",
+                        R"code(Horizontal position of the crop in image coordinates (0.0 - 1.0).)code",
+                        0.5f, true)
+        .AddOptionalArg("crop_pos_y",
+                        R"code(Vertical position of the crop in image coordinates (0.0 - 1.0).)code",
+                        0.5f, true)
+        .AddOptionalArg("mirror",
+                        R"code(Mask for horizontal flip.
 - `0` - do not perform horizontal flip for this image
 - `1` - perform horizontal flip for this image.
 )code", 0, true)
-  .AddOptionalArg("image_type",
-        R"code(The color space of input and output image.)code", DALI_RGB)
-  .AddArg("mean",
-      R"code(Mean pixel values for image normalization.)code",
-      DALI_FLOAT_VEC)
-  .AddArg("std",
-      R"code(Standard deviation values for image normalization.)code",
-      DALI_FLOAT_VEC)
-  .AddArg("crop",
-      R"code(Size of the cropped image. If only a single value `c` is provided,
+        .AddOptionalArg("image_type",
+                        R"code(The color space of input and output image.)code", DALI_RGB)
+        .AddArg("mean",
+                R"code(Mean pixel values for image normalization.)code",
+                DALI_FLOAT_VEC)
+        .AddArg("std",
+                R"code(Standard deviation values for image normalization.)code",
+                DALI_FLOAT_VEC)
+        .AddArg("crop",
+                R"code(Size of the cropped image. If only a single value `c` is provided,
 the resulting crop will be square with size `(c,c)`)code",
-      DALI_INT_VEC);
+                DALI_INT_VEC);
+
+
+// Crop, mirror, mean sub, stddev div, NHWC->NCHW, Npp8u->fp32
+template<typename Out>
+void CropMirrorNormalizePermuteKernel(
+    const int C,
+    const int H,
+    const int W,
+    const bool pad,
+    const int mirror_image,
+    const float* mean,
+    const float* inv_std,
+    const uint8* input_ptr,
+    const int in_step,
+    DALITensorLayout layout,
+    Out* output_ptr) {
+  const int pad_C = C;
+  const int nStride = pad_C*H*W;
+
+  const int a = mirror_image? (W - 1) * C : 0;
+  const int b = mirror_image? -C : C;
+  if (layout == DALI_NCHW) {
+    // Coalesced writes
+    for (int c=0; c < C; ++c) {
+      for (int h=0; h < H; ++h) {
+        for (int w=0; w < W; ++w) {
+          const int in_idx = a + c + b * w + in_step*h;   // HWC
+          const int out_idx = (c*H + h)*W + w;            // CHW
+
+          output_ptr[out_idx] = static_cast<Out>(
+              (static_cast<float>(input_ptr[in_idx])-mean[c]) * inv_std[c]);
```

Comment:
As `mean` and `inv_std` are already floats, the `static_cast<float>` is unnecessary.

Reply:
accepted